blog.md

```diff
@@ -0,0 +1,63 @@
+
+```scala
+def returnInt(x : Int): Int = x
+def returnLong(x : Long): Long = x
+def returnBoolean(x : Boolean): Boolean = x
+```
+
+
+`returnInt` is a function that accepts and returns a value of type `Int` whereas `returnLong` accepts and returns a type `Long` value and finally `returnBoolean`, a `Boolean` type of value.
+
+Although they each return a different **type** of value, they share the same pattern: return the same value that was used as its argument. Since they do the same thing fundamentally, we could write a single function that does that. In other words, we could abstract over this pattern.
+
+````scala
+def identity[A](a: A): A = a
+````
+
+`identity[A]` function is such a function and is called a generic function. The square bracket with A letter in it is the type parameter, syntax of generic function in Scala. When we are generalizing, we tend to use letter A, B, C by convention.
+
+Now, to return an `Int`, we can pass in an `Int` value to the identity function like this, `identity[Int](10)`. When you run this, the output will be a value of 10 which is the same as running `returnInt(10)`. By the same logic, to return a `true` value, we simply write `identity[Boolean](true)`.
+
+Also, you could have as many generic types, also known as type parameters, in a function as you like. For example, there are three different types in `tuple3[A, B, C]` below
+
+````scala
+def tuple3[A, B, C](a: A, b: B, c: C): (A, B, C) = (a, b, c)
+````
+
+So, what is a type? You can think of a type as a set of values. For concrete types like `Boolean`, we know it is a set with just two values: true and false.(draw circle) Earlier, we mentioned that A is a generic type which means we don’t know anything about the possible values lie in set A. (draw circle)
+
+You might be thinking why is having a generic type in a function a good thing? Let’s go back to the `identity[A]` function,  remember how it can take in and return an `Int`, `Long` and `Boolean`. Actually, it can do more than that, it can take in and return **any** type of values due to having a generic type. Compared it to `returnInt`, which can only take in and return a concrete `Int` value, `identity[A]` is a lot more flexible and powerful.
+
+In short, by having a generic function, we can reduce code duplication and achieve polymorphism.
+
+In [day-1-step-1-abstraction](https://github.com/lsug/advent-of-code-typelevel/tree/day-1-step-1-abstraction) branch, there is an exercise called Abstraction. Please try to make every function generic and see if you can pass all the tests.
+
+### Type Class
+
+We’ve seen both type and generic function, let’s go into typeclass. Before bombarding you with lots of definitions of what a typeclass is, we thought it would be good to give you a break and go back to the spacecraft. Remember the `fuel` function, let’s try to rewrite the function and make it generic.
```

Comment:
It would be good to add the generic code snippet here.  e.g.

Remember the `fuel` function.  Let’s try to rewrite the function and make it generic:

```scala
def fuel[A](mass: A): A = mass / 3 - 2
```